### PR TITLE
ANX3SI-373: Dependency updated and Test written + Text Json implement…

### PR DESCRIPTION
**Pull request by anx-aperuzzi**
Dependency updated and Test written + Text Json implementation + CHANGELOG

### CODE REVIEW

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 1]
```diff
@@ -15,49 +17,51 @@ namespace Anexia.BDP.CloudLog
...
+        {
+            PropertyNameCaseInsensitive = true
+        };
+
+        private readonly string _index;
+        private readonly string _token;
+        private readonly string _caFile;
+        private readonly string _certFile;
+        private readonly string _keyFile;
+        private readonly string _keyPassword;
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs]: Those fields should not be necessary any more.

File: src/Anexia.BDP.CloudLog/Anexia.BDP.CloudLog.csproj [hunk 2]
```diff
@@ -13,12 +12,14 @@
     <RepositoryUrl>https://github.com/anexia-it/dotnet-cloudlog</RepositoryUrl>
     <PackageProjectUrl>https://www.anexia-it.com</PackageProjectUrl>
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Anexia.BDP.CloudLog.csproj]: New website URL is: https://www.anexia.com

File: src/Anexia.BDP.CloudLog/Anexia.BDP.CloudLog.csproj [hunk 3]
```diff
@@ -13,12 +12,14 @@
...
     <PackageTags>cloudlog logging</PackageTags>
+    <LangVersion>9</LangVersion>
+    <PackageVersion>1.0.0</PackageVersion>
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Anexia.BDP.CloudLog.csproj]: Can we get the version from an environment variable (e.g. `PACKAGE_VERSION`) instead, with a default value of `0.0.0`? This is needed to implement a GitHub Action that releases the Nuget package when a Release is drafted on Github.

File: src/Anexia.BDP.CloudLog/Anexia.BDP.CloudLog.csproj [hunk 4]
```diff
@@ -13,12 +12,14 @@
...
+    <TargetFrameworks>net5.0;net6.0;net7.0</TargetFrameworks>
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Confluent.Kafka" Version="0.11.5" />
-    <PackageReference Include="Newtonsoft.Json" Version="11.0.2" />
     <PackageReference Include="System.Net.Http" Version="4.3.4" />
+    <PackageReference Include="System.Text.Json" Version="7.0.0" />
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Anexia.BDP.CloudLog.csproj]: Do we need to pin the exact versions here?

File: Anexia.BDP.CloudLog.Test/ClientTest.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            });
+            await clientMoq.PushEvents(new[] { eventString });
+            Assert.Contains("message", result.Records[0].Keys);
+            Assert.Contains("timestamp", result.Records[0].Keys);
+            Assert.Contains("cloudlog_client_type", result.Records[0].Keys);
+            Assert.Contains("cloudlog_source_host", result.Records[0].Keys);
+            Assert.Equal(dataToCheck, result.Records[0].Keys.Count());
+        }
+    }
+}
```
Comment (beachmachine) [inline Anexia.BDP.CloudLog.Test/ClientTest.cs]: Make sure there is always a newline at the end of the file.

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+      - name: Setup .NET Core SDK ${{ matrix.dotnet-version }}
+        uses: actions/setup-dotnet@v1
+        with:
+          dotnet-version: ${{ matrix.dotnet-version }}
+      - name: Install dependencies
+        run: dotnet restore
+      - name: Build
+        run: dotnet build --configuration Release --no-restore
+      - name: Test
+        run: dotnet test --filter Category="GitHub" --no-restore --verbosity normal
```
Comment (beachmachine) [inline .github/workflows/test.yml]: Make sure there is always a newline at the end of the file.

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 7]
```diff
@@ -181,106 +168,53 @@ private void SendOverHttp(string events)
...
-                config.Add("ssl.key.password", keyPassword);
+                throw new ArgumentException("mssing token");
             }
-
-            producer = new Producer<string, string>(config, new StringSerializer(Encoding.UTF8), new StringSerializer(Encoding.UTF8));
         }
-
-
     }
 }
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs]: Make sure there is always a newline at the end of the file.

File: src/Anexia.BDP.CloudLog/Models/PushEventPayload.cs [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        public PushEventPayload(ImmutableDictionary<string, JsonNode>[] records)
+        {
+            Records = records;
+        }
+
+        [JsonPropertyName("records")] public ImmutableDictionary<string, JsonNode>[] Records { get; }
+
+        public static implicit operator PushEventPayload(ImmutableDictionary<string, JsonNode>[] value) => new(value);
+    }
+}
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Models/PushEventPayload.cs]: Make sure there is always a newline at the end of the file.

File: src/Anexia.BDP.CloudLog/Models/PushEventPayload.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
+// // --------------------------------------------------------------------------------------------
+// //  <copyright file = "PushEventPayload.cs" company = "ANEXIA® Internetdienstleistungs GmbH">
+// //  Copyright (c) ANEXIA® Internetdienstleistungs GmbH.All rights reserved.
+// //  </copyright>
+// // --------------------------------------------------------------------------------------------
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Models/PushEventPayload.cs]: Remove this header.

File: Anexia.BDP.CloudLog.Test/ClientTest.cs [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Anexia.BDP.CloudLog.Models;
+using Moq;
+using Moq.Protected;
+using Xunit;
+
+namespace Anexia.BDP.CloudLog.Test
+{
+    [Trait("Category", "GitHub")]
```
Comment (beachmachine) [inline Anexia.BDP.CloudLog.Test/ClientTest.cs]: `[Trait("CloudLog", "Client")]` seems to be a better fit.

File: Anexia.BDP.CloudLog.Test/ClientTest.cs [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        [InlineData(
+            "{\"message\":\"Something\",\"timestamp\":\"1669816693\", \"cloudlog_client_type\":\"SomeType\", \"cloudlog_source_host\": \"SomeHost\"}",
+            4)]
+        [InlineData(
+            "{\"message\":\"Something\", \"cloudlog_client_type\":\"SomeType\", \"cloudlog_source_host\": \"SomeHost\"}",
+            4)]
+        [InlineData(
+            "{\"message\":\"Something\", \"cloudlog_client_type\":\"SomeType\", \"cloudlog_source_host\": \"SomeHost\", \"some_other_data\": \"SomeExtensiveData\", \"some_more_numbers\": 1541864}",
+            6)]
+        public async Task PushEvent(string eventString, int dataToCheck)
```
Comment (beachmachine) [inline Anexia.BDP.CloudLog.Test/ClientTest.cs]: `expectedKeyCount` seems to be more clear than `dataToCheck`.

File: Anexia.BDP.CloudLog.Test/ClientTest.cs [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        [InlineData(
+            "{\"message\":\"Something\",\"timestamp\":\"1669816693\", \"cloudlog_client_type\":\"SomeType\", \"cloudlog_source_host\": \"SomeHost\"}",
+            4)]
+        [InlineData(
+            "{\"message\":\"Something\", \"cloudlog_client_type\":\"SomeType\", \"cloudlog_source_host\": \"SomeHost\"}",
+            4)]
+        [InlineData(
+            "{\"message\":\"Something\", \"cloudlog_client_type\":\"SomeType\", \"cloudlog_source_host\": \"SomeHost\", \"some_other_data\": \"SomeExtensiveData\", \"some_more_numbers\": 1541864}",
+            6)]
+        public async Task PushEvents(string eventString, int dataToCheck)
```
Comment (beachmachine) [inline Anexia.BDP.CloudLog.Test/ClientTest.cs]: `expectedKeyCount` seems to be more clear than `dataToCheck`.

**Review by beachmachine — CHANGES_REQUESTED** [review]

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 13]
```diff
@@ -15,49 +17,47 @@ namespace Anexia.BDP.CloudLog
...
-        private string clientType;
-        private bool isHttp;
+        private static JsonSerializerOptions _jsonSerializerOptions = new()
+        {
+            PropertyNameCaseInsensitive = true
+        };
+
+        private readonly string _index;
+        private readonly string _token;
+        private readonly bool _isHttp;
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs]: Remove `_isHttp`, as it is unused.

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 14]
```diff
@@ -15,49 +17,47 @@ namespace Anexia.BDP.CloudLog
...
 
-        private string index;
-        private string token;
-        private string caFile;
-        private string certFile;
-        private string keyFile;
-        private string keyPassword;
-        private string clientType;
-        private bool isHttp;
+        private static JsonSerializerOptions _jsonSerializerOptions = new()
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs]: You can mark this field as `readonly`.

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 15]
```diff
@@ -15,49 +17,47 @@ namespace Anexia.BDP.CloudLog
     /// </summary>
     public class Client
     {
-        private string brokers = "kafka0401.bdp.anexia-it.com:8443";
-        private string api = "https://api0401.bdp.anexia-it.com";
+        private const string api = "https://api0401.bdp.anexia-it.com";
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs]: Constants should use UpperCamelCase. Rename to `Api`.

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 16]
```diff
@@ -15,49 +17,47 @@ namespace Anexia.BDP.CloudLog
...
-        private bool isHttp;
+        private static JsonSerializerOptions _jsonSerializerOptions = new()
+        {
+            PropertyNameCaseInsensitive = true
+        };
+
+        private readonly string _index;
+        private readonly string _token;
+        private readonly bool _isHttp;
         private long queue = 0;
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs]: I do not see the need for this counter any more. Remove if possible. Else, private instance fields should start with `_`. Rename to `_queue`.

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 17]
```diff
@@ -15,49 +17,47 @@ namespace Anexia.BDP.CloudLog
...
 
-        private Producer<string, string> producer;
+        private string ClientType;
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs]: As we only support HTTP, I do not see the need for this field any more. Remove if possible. Else, private instance fields should start with `_` and should be in lowerCamelCase. Rename to `_clientType`.

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 18]
```diff
@@ -15,49 +17,47 @@ namespace Anexia.BDP.CloudLog
...
+
         private HttpClient httpClient;
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs]: Private instance fields should start with `_`. Rename to `_httpClient`.

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 19]
```diff
@@ -15,49 +17,47 @@ namespace Anexia.BDP.CloudLog
...
 
+        private readonly Func<long> createTimeStamp =
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs]: Either keep this an lambda expression and rename to `_createTimeStamp` or convert it to a proper method and rename to `CreateTimeStamp`.

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 20]
```diff
@@ -181,106 +164,53 @@ private void SendOverHttp(string events)
...
 
         /// <summary>
         ///     Check http configuration
         /// </summary>
         private void CheckHttpConfiguration()
         {
-            if (String.IsNullOrEmpty(index))
+            if (String.IsNullOrEmpty(_index))
             {
                 throw new ArgumentException("mssing index");
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs]: Typo. Set exception message to `"index is missing"`.

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 21]
```diff
@@ -181,106 +164,53 @@ private void SendOverHttp(string events)
...
-                { "ssl.key.location", keyFile },
-                { "request.required.acks", "all" },
-                { "compression.codec", "gzip" },
-                { "retries", 10 },
-
-            };
-            if (!string.IsNullOrWhiteSpace(keyPassword))
-            {
-                config.Add("ssl.key.password", keyPassword);
+                throw new ArgumentException("mssing token");
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs]: Typo. Set exception message to `"token is missing"`.

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 22]
```diff
@@ -98,61 +91,49 @@ public void SetClientType(string clientType)
...
-                JArray array = new JArray();
-                foreach (string evt in events)
-                {
-                    var finalEvt = AddMetadata(evt, timestamp);
-                    array.Add(finalEvt);
-                }
-                SendOverHttp(array.ToString());
-            }
-            else
+            long timestamp = createTimeStamp();
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs]: We can use `var` here.

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 23]
```diff
@@ -98,61 +91,49 @@ public void SetClientType(string clientType)
...
+            ImmutableDictionary<string, JsonNode>[] result =
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs]: We can use `var` here.

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 24]
```diff
@@ -98,61 +91,49 @@ public void SetClientType(string clientType)
...
+        /// <summary>
+        ///     Sends an event to CloudLog
+        /// </summary>
+        /// <param name="evt">
+        ///     Event
+        /// </param>
+        /// <returns>Task of post async function</returns>
+        public Task PushEvent(string evt)
+        {
+            long timestamp = createTimeStamp();
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs]: We can use `var` here.

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 25]
```diff
@@ -98,61 +91,49 @@ public void SetClientType(string clientType)
...
+                result[i] = AddMetadata(ref eventObj, ref timestamp);
             }
 
+            return SendOverHttp(result);
         }
 
         /// <summary>
         ///     Flush events
         /// </summary>
         public void Flush()
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs]: I do not see the need for this method any more. Remove if possible.

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 26]
```diff
@@ -89,7 +82,7 @@ public Client(string index, string token)
         /// </param>
         public void SetClientType(string clientType)
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs:79]: Remove method if possible.

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 27]
```diff
@@ -15,49 +17,47 @@ namespace Anexia.BDP.CloudLog
...
+        private string ClientType;
+
         private HttpClient httpClient;
 
+        private readonly Func<long> createTimeStamp =
+            () => (long)DateTime.Now.Subtract(DateTime.MinValue.AddYears(1969)).TotalMilliseconds;
+
         /// <summary>
-        ///     Creates a new client instance (kafka)
+        ///     Creates a new client instance (http)
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs]: I do not think we need to mention `"(http)"` here, as this is the only thing we support.

File: src/Anexia.BDP.CloudLog/Client.cs [hunk 28]
```diff
@@ -15,49 +17,47 @@ namespace Anexia.BDP.CloudLog
...
+            _token = token;
+            _isHttp = true;
+            ClientType = "dotnet-client-http";
+            CheckHttpConfiguration();
 
+            httpClient = client;
+            httpClient.DefaultRequestHeaders.TryAddWithoutValidation("Authorization", token);
         }
 
         /// <summary>
```
Comment (beachmachine) [inline src/Anexia.BDP.CloudLog/Client.cs:59]: I do not think we need to mention `"(http)"` here, as this is the only thing we support.

File: src/Anexia.BDP.CloudLog/Anexia.BDP.CloudLog.csproj (hunk 3 above)
Comment (beachmachine) [reply src/Anexia.BDP.CloudLog/Anexia.BDP.CloudLog.csproj]: Why was this resolved? Nothing has changed.

File: Anexia.BDP.CloudLog.Test/ClientTest.cs [hunk 29]
```diff
@@ -0,0 +1,87 @@
...
+using Moq;
+using Moq.Protected;
+using Xunit;
+
+namespace Anexia.BDP.CloudLog.Test
+{
+    [Trait("CloudLog", "Client")]
+    public class ClientTest
+    {
+        private const string testContent = "test content";
```
Comment (beachmachine) [inline Anexia.BDP.CloudLog.Test/ClientTest.cs]: Constants should use UpperCamelCase. Rename to `TestContent`.

**Review by beachmachine — CHANGES_REQUESTED** [review]

File: README.md [hunk 30]
```diff
@@ -1,2 +1,94 @@
...
+
+`dotnet-cloudlog` is a client library for Anexia CloudLog. It provides a simple API for sending events to a CloudLog
+index directly from your .NET application.
+
+**Note:** Usually it is considered best-practice to write rotating log-files to the filesystem, and send those logs to
+CloudLog via `Filebeat`.
+
+# Install
+
+With a correctly set up .NET SDK, run in `PowerShell``:
```
Comment (beachmachine) [inline README.md]: There are double backticks at the end of `"PowerShell"`. Remove on of them.

File: src/Anexia.BDP.CloudLog/Anexia.BDP.CloudLog.csproj (hunk 3 above)
Comment (anx-aperuzzi) [reply src/Anexia.BDP.CloudLog/Anexia.BDP.CloudLog.csproj]: It should be already been changed

File: src/Anexia.BDP.CloudLog/Anexia.BDP.CloudLog.csproj (hunk 4 above)
Comment (anx-aperuzzi) [reply src/Anexia.BDP.CloudLog/Anexia.BDP.CloudLog.csproj]: yes

File: src/Anexia.BDP.CloudLog/Client.cs (hunk 26 above)
Comment (anx-aperuzzi) [reply src/Anexia.BDP.CloudLog/Client.cs:79]: Not possible it's used by Serilog.Sinks.Cloudlog

Comment (beachmachine) [reply src/Anexia.BDP.CloudLog/Client.cs:79]: We will need to release a new version of `Serilog.Sinks.Cloudlog` anyway, so we remove the call to `SetClientType` there too.

Comment (beachmachine) [reply src/Anexia.BDP.CloudLog/Client.cs:79]: As discussed, we keep the `SetClientType` method as it is used by the `Serilog.Sinks.Cloudlog` library so set an all different client type.

**Review by beachmachine — APPROVED** [review]

**Pull request merged**